flyway/sql/V18__registration_funnel_1.sql

```diff
@@ -0,0 +1,49 @@
+-- for some columns we never let the caller control the contents
+CREATE OR REPLACE FUNCTION mappa.override_registration_funnel_columns()
+RETURNS TRIGGER AS $$
+BEGIN
+  IF NEW.uuid IS NOT NULL THEN
+    RAISE EXCEPTION 'You must not send uuid field';
+  ELSE
+    NEW.uuid = mappa.uuid_generate_v4();
+  END IF;
+  IF NEW.created_at IS NOT NULL THEN
+    RAISE EXCEPTION 'You must not send created_at field';
+  ELSE
+    NEW.created_at = now();
+  END IF;
+  IF NEW.email IS NULL THEN
+    RAISE EXCEPTION 'Must provide an email address';
+  END IF;
+  IF NEW.survey_results IS NULL THEN
+    RAISE EXCEPTION 'You must submit survey results';
+  END IF;
+  IF NEW.ip_address IS NULL THEN
+    RAISE EXCEPTION 'Client IP is required';
+  END IF;
+  IF NEW.registered IS NOT NULL THEN
+    RAISE EXCEPTION 'You must not send the registered field';
+  ELSE
+    NEW.registered = FALSE;
+  END IF;
+
+  RETURN NEW;
+END;
+$$ language 'plpgsql';
+
+CREATE TABLE mappa.registration_funnel_1 (
+  uuid           UUID        UNIQUE,
+  created_at     TIMESTAMPTZ,
+  email          NAME,
```

Comment:
Where can I find the definition of the NAME type?

Reply:
Aha, found it. Name is a 64 byte charactrer type used for internal object names: https://www.postgresql.org/docs/9.6/static/datatype-character.html

Since email addresses may be 254 characters, I think we should use a different type. Some ideas in [this post](https://dba.stackexchange.com/questions/68266/what-is-the-best-way-to-store-an-email-address-in-postgresql). citext seems to be the simplest thing.

Reply:
@statik hi! I used this [NAME](https://www.postgresql.org/docs/9.1/static/datatype-character.html ) data type since it was used with earlier migrations in the project. Although I think this might raise a good discussion on its usage. There may be better options such as VARCHAR(320). What do you think?

Reply:
hah! It looks like I'm the one that used NAME at first, probably copy and paste from an example without thinking about it very carefully. I think it would make sense to migrate from `NAME` to `citext` for any email columns.